### ISSUE TITLE
[propose feature] Adding sparsemax

### ISSUE THREAD

**Issue by nicola-decao** [pytorch/pytorch#14923]
## 🚀 Feature
I would like to add the function and `torch.nn.Module` for sparsemax as described in https://arxiv.org/abs/1602.02068

## Motivation

It is a function that I implemented in plain pytorch and I would like to share it with the community and enrich the library.

## Pitch

I have already have a piece of code that does that with both forward and backward. It should work for any batch dimensions and it applies sparsemax on the dimension `dim`. However, I think it should be revisited as the standard required by the library.

```
import torch

class SparsemaxFunction(torch.autograd.Function):

    @staticmethod
    def forward(self, z, dim=-1):
        z = z.transpose(dim, -1)
        sorted_z, _ = z.sort(-1, descending=True)
        k_z = (1 + torch.arange(1., 1 + z.shape[-1]) * sorted_z > sorted_z.cumsum(-1)).sum(-1, keepdim=True) + 1
        mask = (k_z >= torch.arange(1, z.shape[-1] + 1).repeat(z.shape[:-1] + (1, )))
        t_z = ((mask.float() * sorted_z).sum(-1, keepdim=True) - 1) / k_z.float()
        out = torch.relu(z - t_z)
        self.save_for_backward(out, torch.tensor(dim))
        return out.transpose(dim, -1)

    @staticmethod
    def backward(self, grad_output):
        dim = self.saved_tensors[1]
        s = (self.saved_tensors[0] > 0).float()
        v = (s * grad_output).sum(dim, keepdim=True) / s.sum(dim, keepdim=True)
        grad = s * (grad_output - v)
        return grad
    
sparsemax = SparsemaxFunction.apply
    
class Sparsemax(torch.nn.Module):
    
    def __init__(self):
        super().__init__()
        
    def forward(self, z):
        return SparsemaxFunction.apply(z)
```

## Alternatives

This function might be implemented directly in C++ since it is quite simple.

cc @ezyang @gchanan @zou3519

**Comment by around1991** [comment]
This code is slightly wrong out of the box because ``argmax`` isn't guaranteed to return the highest index of the max if multiple indices are tied. Concretely, on my setup it works fine for CPU but gives the wrong result for GPU (``k_z`` is instead a tensor of all 0's).

**Comment by nicola-decao** [comment]
> This code is slightly wrong out of the box because `argmax` isn't guaranteed to return the highest index of the max if multiple indices are tied. Concretely, on my setup it works fine for CPU but gives the wrong result for GPU (`k_z` is instead a tensor of all 0's). `k_z = (...).sum(-1, keepdim=True)` works instead.

@around1991  You are right! Thanks for pointing it out. I will correct this in the proposal.

**Comment by around1991** [comment]
Also another issue: it won't handle -inf correctly, because 0*-inf gives nan in the forward pass. I think you need to multiply `mask` by a version of `k_z` where you use torch.where to get rid of the -inf to handle this properly.

**Comment by ntenenz** [comment]
Is there still interest in this feature? I'm happy to take a stab at it but would hate to waste the energy if there's no potential for a future merge.

**Comment by cpuhrsch** [comment]
Flagging this for triage review so we can decide how to proceed with this.
